OpenRA.Mods.Common/Traits/World/ResourceLayer.cs

```diff
@@ -17,35 +18,52 @@
 namespace OpenRA.Mods.Common.Traits
 {
 	[Desc("Attach this to the world actor.", "Order of the layers defines the Z sorting.")]
-	public class ResourceLayerInfo : TraitInfo<ResourceLayer>, Requires<ResourceTypeInfo> { }
+	public class ResourceLayerInfo : ITraitInfo, Requires<ResourceTypeInfo>, Requires<BuildingInfluenceInfo>
+	{
+		public virtual object Create(ActorInitializer init) { return new ResourceLayer(init.Self); }
+	}
 
 	public class ResourceLayer : IRenderOverlay, IWorldLoaded, ITickRender
 	{
 		static readonly CellContents EmptyCell = new CellContents();
 
-		World world;
+		readonly World world;
+		readonly BuildingInfluence buildingInfluence;
+		readonly HashSet<CPos> dirty = new HashSet<CPos>();
+		readonly Dictionary<PaletteReference, TerrainSpriteLayer> spriteLayers = new Dictionary<PaletteReference, TerrainSpriteLayer>();
 
-		BuildingInfluence buildingInfluence;
+		protected readonly CellLayer<CellContents> Content;
+		protected readonly CellLayer<CellContents> RenderContent;
 
-		protected CellLayer<CellContents> content;
-		protected CellLayer<CellContents> render;
-		List<CPos> dirty;
+		public ResourceLayer(Actor self)
+		{
+			world = self.World;
+			buildingInfluence = self.Trait<BuildingInfluence>();
 
-		public void Render(WorldRenderer wr)
+			Content = new CellLayer<CellContents>(world.Map);
+			RenderContent = new CellLayer<CellContents>(world.Map);
+
+			RenderContent.CellEntryChanged += UpdateSpriteLayers;
+		}
+
+		void UpdateSpriteLayers(CPos cell)
 		{
-			var shroudObscured = world.ShroudObscuresTest;
-			foreach (var uv in wr.Viewport.VisibleCellsInsideBounds.MapCoords)
+			var resource = RenderContent[cell];
+			foreach (var kv in spriteLayers)
 			{
-				if (shroudObscured(uv))
-					continue;
-
-				var c = render[uv];
-				if (c.Sprite != null)
-					new SpriteRenderable(c.Sprite, wr.World.Map.CenterOfCell(uv.ToCPos(world.Map)),
-						WVec.Zero, -511, c.Type.Palette, 1f, true).Render(wr); // TODO ZOffset is ignored
+				if (resource.Sprite != null && resource.Type.Palette == kv.Key)
```

Comment:
That would be good in case somebody else comes to the same conclusion I just did.